bench/src/bulk.rs

```diff
@@ -80,34 +136,143 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
     client_config
         .add_certificate_authority(server_cert)
         .unwrap();
+    let mut client_config = client_config.build();
+    client_config.transport = Arc::new(transport_config());
+
     let quinn::NewConnection { connection, .. } = endpoint
-        .connect_with(client_config.build(), &server_addr, "localhost")
+        .connect_with(client_config, &server_addr, "localhost")
         .unwrap()
         .await
         .context("unable to connect")?;
     trace!("connected");
 
-    let mut stream = connection
-        .open_uni()
-        .await
-        .context("failed to open stream")?;
-    const DATA: &[u8] = &[0xAB; 1024 * 1024];
     let start = Instant::now();
-    for _ in 0..1024 {
-        stream
-            .write_all(DATA)
-            .await
-            .context("failed sending data")?;
+
+    let connection = Arc::new(connection);
+    let get_connection = || connection.clone();
+
+    let mut ops = futures::stream::iter((0..config.streams).map(|_| async move {
+        send_data_on_stream(get_connection(), config.stream_size_mb).await
+    }))
+    .buffer_unordered(config.max_streams);
+
+    let mut total_size = 0;
+    let mut duration_hist: Histogram<u64> = Histogram::new(3).unwrap();
+    let mut throughput_hist: Histogram<u64> = Histogram::new(3).unwrap();
+
+    while let Some(result) = ops.next().await {
+        info!("stream finished: {:?}", result);
+        let result = result.unwrap();
+        total_size += result.size;
+
+        duration_hist
+            .record(result.duration.as_millis() as u64)
+            .unwrap();
+        throughput_hist.record(result.throughput as u64).unwrap();
     }
-    stream.finish().await.context("failed finishing stream")?;
-    println!("sent {} bytes in {:?}", 1024 * DATA.len(), start.elapsed());
+
+    let dt = start.elapsed();
+    println!("Overall stats:\n");
+    println!(
+        "Sent {} bytes on {} streams in {:4.2?} ({:.2} MiB/s)\n",
+        total_size,
+        config.streams,
+        dt,
+        throughput_bps(dt, total_size as u64) / 1024.0 / 1024.0
+    );
+
+    println!("Stream metrics:\n");
+
+    println!(
+        "AVG : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.mean() as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.mean() as u64)
+    );
+    println!(
+        "P0  : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
```

Comment:
I find the ordering a bit counterintuitive here, I guess because percentiles are more usually used for latencies, so I expect the highest percentile to be worst and the lowest percentile to be best. You probably have more experience with throughput distribution representations than me, though...

Reply:
I don't think there is a right or wrong here. Both approaches work. I actually want to add first-byte latency as another column, and there P0 is the best value and P100 the worst 🤷‍♂️

Reply:
I was confused about the example output because it seemed to suggest that the lowest-throughput stream had the shortest duration. Not sure how to do better without making it massively less compact, though.

Reply:
I  agree. If one interprets it as "these are the values of individual streams", it gets a bit confusing. It should be clear that those are not, because we have far more streams, but 🤷‍♂️

One thing that can be changed is to make it a real table, and move the throughput / duration labels to a separate row. Not sure if this is clearer or not however.

Individual stream metrics are found in the the trace:

```
Dec 27 12:00:39.056  INFO bulk: stream finished: Ok(SendResult { duration: 11.809008s, size: 10485760, throughput: 887945.8799587569 })
```

Reply:
Here's the table format. I think I slightly prefer it:

```
Stream metrics:

      │  Throughput   │ Duration
──────┼───────────────┼──────────
 AVG  │  155.43 MiB/s │  391.00ms
 P0   │    0.84 MiB/s │   39.00ms
 P10  │   50.56 MiB/s │   40.00ms
 P50  │  131.25 MiB/s │   76.00ms
 P90  │  249.87 MiB/s │  195.00ms
 P100 │  255.25 MiB/s │    11.97s
```

Reply:
Yeah, I like the table format, too.